src/esp/assets/PTexMeshData.cpp

```diff
@@ -469,13 +492,16 @@ void PTexMeshData::parsePLY(const std::string& filename,
         break;
       } else {
         // Something unrecognised
-        ASSERT(false);
+        CORRADE_ASSERT(false, "Error: something is unrecognised.", );
       }
     }
 
     // Check things make sense.
-    ASSERT(numVertices > 0);
-    ASSERT(positionDimensions > 0);
+    CORRADE_ASSERT(numVertices > 0,
+                   "Error: number of vertices is not greater than 0", );
+    CORRADE_ASSERT(
+        positionDimensions > 0,
+        "Error: the dimensions of the position is not greater than 0", );
```

Comment:
Is the `Error: ` prefix really needed? :) In every case it's the last thing you see before it aborts.

In Magnum itself I'm instead prefixing every assert with class/function name to make it easier to figure out where it blew up. In this case it would be e.g. `PTexMeshData::parsePly(): the dimensions...`.

(Also, I'm planning to add configurable message prefixes / coloring in the near future.)

Reply:
Good suggestion. I used to do what you recommended as well, and then I stopped to do so since the ASSERT, which used LOG, would output the filename and line number.

@mosra: Another question to you about the CORRADE_ASSERT, it seems it would generate a run-time error instead of exiting the program (like e.g., exit(-1) does)?

Reply:
A bit more background on this: there's `CORRADE_INTERNAL_ASSERT()` which doesn't have a message but OTOH prints file/line info. That one is designed for internal assertions which shouldn't ever fire as a result of bad API usage by the library user -- and since the internal error messages would be in most cases just `"internal state that's too complex to explain in a single sentence is totally messed up, sorry for your bad experience"` I decided to not have them there.

OTOH I thought showing file/line info in `CORRADE_ASSERT()` would be mostly useless for the user as they rather need to know what function *name* caused everything to blow up (as opposed to knowing it's line 2765 of AbstractTexture.cpp, for example). So for those there's just a human-readable formatted message.

Basically when deciding which of them to use, I think whether the assertion can happen rather from bad usage on the API user side (in which case it's more important to show *why* is it wrong), or it's just an internal sanity check that shouldn't hit when just passing bad parameters to a function (in which case file/line info is better). ... Hope those design decisions still make sense now :)

> it seems it would generate a run-time error instead of exiting the program (like e.g., exit(-1) does)?

Yes, it calls `std::abort()` instead of `std::exit()` -- reason is that the former generates a proper backtrace (or enters a debugger on some systems), which is what you usually want when an assert hits. The `exit()` would just exit without giving you any chance to know from where it was fired.

Reply:
OK, I see your point on the design decisions. I totally understand you. 

To decide which one to use, we need to think about the purpose of the error message (to expedite debugging or FYI only), and who is the audience (the developer or the user of the system).

Though it is pretty vague in some cases, it is not hard to determine.

I will go over all the places I changed, and update accordingly.

Reply:
Oh, you used `std::abort()`. That is why.

Yes, the returned backtrace is rather useful sometimes for the developer. 

Again, if the audience is a user of our simulator, in my humble opinion, it is more comfortable for her to see the program is terminated normally (exit()) instead of abnormally (abort()).

Reply:
But it didn't terminate normally -- It terminated due to a failed assertion!

Reply:
> But it didn't terminate normally -- It terminated due to a failed assertion!

Yes, indeed, that is actually what happened. I agree, but that is just from the developer's perspective.

But from a user's view, maybe it is more comfortable if the system can "pretend" to quit "normally",meaning it exits without a run-time error (otherwise, there would be an annoying pop-up dialog-box on mac e.g.,).

Reply:
I definitely don't think people want things to crash with no indication that anything went wrong.  It seems very for them to figure out that something is broken if we don't inform them in a fairly loud way.

Reply:
If you want to `exit()` with a message, use `Fatal{} << blah;`. It exits the "normal" way, it's what `Utility::Arguments` do when you pass a wrong thing on the command line. Asserts are loud by definition and yes, are meant for developers. If a non-developer hits an assert, it's the app developers fault this wasn't checked and handled gracefully.

Reply:
Hmmm.  Maybe something like `FATAL_IF(<blah>, <msg>)` would be useful?  Some of the stuff here is checking if a file exists, so a user could in theory hit that.  The assert style synatx is nice than `if (<blah>) Fatal{} << <msg>;` IMO

Reply:
As I understand it, asserts are macros mainly so there's a possibility to compile them out in optimized builds. Apart from that I'm trying to avoid macros as much as possible, so the non-macroy `if()` is no problem for me.